logical: use implicit txns for each INSERT and DELETE

yuzefovich opened the pull request:
**logical: use implicit txns for each INSERT and DELETE**

This commit introduces another strategy for processing rows. Previously, we would handle a batch of rows within an explicit txn, and this commit makes it so that each row is handled in a separate implicit txn. This strategy is now used by default and can be disabled via newly-added `logical_replication.consumer.use_implicit_txns.enabled` cluster setting. Usage of the implicit txns allows us to hit 1PC down in the KV layer in some cases.

**logical: remove some overhead with explicit txns**

Previously, when using the explicit txns to perform INSERT and DELETE ingestion queries we would use "vanilla" `Txn` method. This method creates a new internal session data on each invocation of the internal executor methods. This commit avoids doing that by supplying the Txn method with the copy of the session data from the session that started the replication stream - that copy will be adjusted to apply some internal executor overrides but otherwise will be unchanged. In other words, even though this commit is not exactly a no-op, it is effectively one.

Epic: CRDB-39063.

Release note: None

yuzefovich committed:
logical: use implicit txns for each INSERT and DELETE

This commit introduces another strategy for processing rows. Previously,
we would handle a batch of rows within an explicit txn, and this commit
makes it so that each row is handled in a separate implicit txn. This
strategy is now used by default and can be disabled via newly-added
`logical_replication.consumer.use_implicit_txns.enabled` cluster
setting. Usage of the implicit txns allows us to hit 1PC down in the KV
layer in some cases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126050)
<!-- Reviewable:end -->

yuzefovich commented:
First commit is #126051.

The benchmarks show some slowdown because we now create more state on each `ExecParsed` call.
```
name                                 old time/op    new time/op    delta
LastWriteWinsInsert/noPrevValue-8       802µs ± 1%     830µs ± 1%  +3.55%  (p=0.000 n=9+8)
LastWriteWinsInsert/withPrevValue-8    1.18ms ± 1%    1.21ms ± 1%  +2.64%  (p=0.000 n=10+10)

name                                 old alloc/op   new alloc/op   delta
LastWriteWinsInsert/noPrevValue-8       177kB ± 4%     180kB ± 4%  +1.32%  (p=0.019 n=9+9)
LastWriteWinsInsert/withPrevValue-8     242kB ± 1%     244kB ± 0%  +0.80%  (p=0.000 n=9+8)

name                                 old allocs/op  new allocs/op  delta
LastWriteWinsInsert/noPrevValue-8       1.08k ± 1%     1.17k ± 1%  +8.72%  (p=0.000 n=8+8)
LastWriteWinsInsert/withPrevValue-8     1.63k ± 0%     1.73k ± 0%  +5.69%  (p=0.000 n=9+9)
```
We can avoid this by splitting `ProcessRow` into two methods.

yuzefovich committed:
logical: remove some overhead with explicit txns

Previously, when using the explicit txns to perform INSERT and DELETE
ingestion queries we would use "vanilla" `Txn` method. This method
creates a new internal session data on each invocation of the internal
executor methods. This commit avoids doing that by supplying the Txn
method with the copy of the session data from the session that started
the replication stream - that copy will be adjusted to apply some
internal executor overrides but otherwise will be unchanged. In other
words, even though this commit is not exactly a no-op, it is effectively
one.

```
name                                             old time/op    new time/op    delta
LastWriteWinsInsert/explicitTxn/noPrevValue-8       812µs ± 2%     758µs ± 2%  -6.61%  (p=0.000 n=19+20)
LastWriteWinsInsert/explicitTxn/withPrevValue-8    1.19ms ± 1%    1.14ms ± 1%  -3.91%  (p=0.000 n=20+20)
LastWriteWinsInsert/implicitTxn/noPrevValue-8       747µs ± 1%     752µs ± 2%  +0.76%  (p=0.018 n=19+20)
LastWriteWinsInsert/implicitTxn/withPrevValue-8    1.14ms ± 2%    1.14ms ± 1%  +0.52%  (p=0.015 n=20+19)

name                                             old alloc/op   new alloc/op   delta
LastWriteWinsInsert/explicitTxn/noPrevValue-8       177kB ± 2%     174kB ± 2%  -1.45%  (p=0.000 n=17+16)
LastWriteWinsInsert/explicitTxn/withPrevValue-8     243kB ± 0%     240kB ± 0%  -1.02%  (p=0.000 n=16+16)
LastWriteWinsInsert/implicitTxn/noPrevValue-8       166kB ± 9%     165kB ± 3%    ~     (p=0.985 n=16+16)
LastWriteWinsInsert/implicitTxn/withPrevValue-8     231kB ± 0%     231kB ± 0%    ~     (p=0.897 n=16+16)

name                                             old allocs/op  new allocs/op  delta
LastWriteWinsInsert/explicitTxn/noPrevValue-8       1.07k ± 2%     0.99k ± 2%  -7.39%  (p=0.000 n=17+16)
LastWriteWinsInsert/explicitTxn/withPrevValue-8     1.63k ± 0%     1.55k ± 0%  -4.83%  (p=0.000 n=17+17)
LastWriteWinsInsert/implicitTxn/noPrevValue-8         974 ±11%       967 ± 1%    ~     (p=0.771 n=16+16)
LastWriteWinsInsert/implicitTxn/withPrevValue-8     1.53k ± 0%     1.53k ± 0%    ~     (p=0.843 n=17+17)
```

Release note: None

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go:
Do we still need this here now that we are setting this in the session data?

stevendanna reviewed (approved):
Left a small question but overall this seems pretty good to me.  Makes sense that we wouldn't see an improvement in the micro-benchmark case. Will be interesting to see if this has any impact on a larger test.

yuzefovich reviewed (commented):
There was initially some regression which I had to fight back in order to get "no difference in overhead" :smiley:

TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126050#-:-O0ADDAW1-xClo57vD1b:b-g1v0gx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go` line 591 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126050#gh-1650348437:-O0ACnk77mLxxcG16JWU:b6iryer) ([raw file](https://github.com/cockroachdb/cockroach/blob/6914507a35add039761f03b6142ed4d03f627f85/pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go#L591)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Do we still need this here now that we are setting this in the session data?
</blockquote></details>

No, it's not needed, removed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126050)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go` line 591 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/126050#gh-1650348437:-O0H6d1F686Ovcz2WnkA:b-ahyhfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6914507a35add039761f03b6142ed4d03f627f85/pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go#L591)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

No, it's not needed, removed.
</blockquote></details>

Turns out that we still need it - https://github.com/cockroachdb/cockroach/pull/126234.


<!-- Sent from Reviewable.io -->
